master/internal/task/allgather.go

```diff
@@ -0,0 +1,40 @@
+package task
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+
+	"github.com/determined-ai/determined/master/internal/task/allgather"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AllGather blocks until `numPeers` with the same `allocationID` are waiting and then returns the
+// data from all those peers. It returns an error in the if the call returns early, without data,
+// for any reason. Only one call may connect per `id`.
+func AllGather(
```

Comment:
i'll probably ticket making `allgather.Join` sync just like this, removing `allgather.Leave`, and probably removing a lot of the code (watchers in particular). but in the spirit of keeping the changes super minimal, i'm gonna keep most of the weirdness